pkg/operator/controller/dnsnameresolver-feature/controller.go

```diff
@@ -0,0 +1,160 @@
+		cache:  mgr.GetCache(),
+		client: mgr.GetClient(),
+		config: config,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	clusterNamePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		expectedName := operatorcontroller.FeatureGateClusterConfigName()
+		actualName := types.NamespacedName{
+			Namespace: o.GetNamespace(),
+			Name:      o.GetName(),
+		}
+		return expectedName == actualName
+	})
+	if err := c.Watch(source.Kind(mgr.GetCache(), &configv1.FeatureGate{}), &handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DNSNameResolverEnabled indicates that the "DNSNameResolver" featuregate is enabled.
+	DNSNameResolverEnabled bool
+
+	FeatureGateName string
+
+	// DependentCaches is a list of caches that are used by Controllers watching DNSNameResolver
+	// resources. The dnsnameresolverfeature controller starts these caches once
+	// the DNSNameResolver CRD has been created.
+	DependentCaches []cache.Cache
+	// DependentControllers is a list of controllers that watch DNSNameResolver
+	// resources. The dnsnameresolverfeature controller starts these controllers once
+	// the DNSNameResolver CRD has been created and the DependentCaches are started.
+	DependentControllers []controller.Controller
+}
+
+// reconciler handles the actual featuregate reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+
+	cache            cache.Cache
+	client           client.Client
+	startCaches      sync.Once
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a FeatureGate and creates or
+// reconciles the DNSNameResolver CRD.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logrus.Info("reconciling", "request", request)
+
+	// Check if the DNSNameResolved feature gate is enabled or not.
+	if !r.config.DNSNameResolverEnabled {
+		return reconcile.Result{}, nil
+	}
+
+	// Check if the name of feature gate matches with the default feature gate.
+	if request.Name != r.config.FeatureGateName {
+		// Return a nil error value to avoid re-triggering the event.
+		logrus.Errorf("skipping unexpected feature gate object %s", request.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// Get the feature gate object.
+	featureGate := configv1.FeatureGate{}
+	if err := r.cache.Get(ctx, request.NamespacedName, &featureGate); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the feature set is either TechPreviewNoUpgrade or CustomNoUpgrade.
+	if featureGate.Spec.FeatureSet != configv1.TechPreviewNoUpgrade && featureGate.Spec.FeatureSet != configv1.CustomNoUpgrade {
```

Comment:
I don't think we should handle both feature sets in one release.  Can you convince me?

Reply:
As mentioned in the comment (https://github.com/openshift/cluster-dns-operator/pull/394#discussion_r1458903735) the DNSNameResolver feature can be enabled by either TechPreviewNoUpgrade or CustomNoUpgrade feature set. We need to create the CRD for the corresponding feature set.